### ISSUE TITLE
Move to_<num_type> from String to Slice

### ISSUE THREAD

**Issue by bew** [crystal-lang/crystal#6389]
While working on a terminfo parsing library, where I manipulate Slice of bytes (`Bytes` type), I needed to convert digits to an actual number.
I realized that all the `to_<num_type>` conversion methods (like `to_i64`, `to_u8`, `to_f`, ..) are in `String`, and I needed to allocate a `String` just to convert the bytes to a number.

I think thoses methods should be available in `Slice` too. The number conversion methods don't need anything from `String` as they directly work with a pointer and a size. They can almost directly be copy/pasted to `Slice`!

I have a working [branch](https://github.com/bew/crystal/tree/to_num-on-slice) with this change, I can make a PR when it's accepted!

Note: the only problem I see here is that those methods will now exists for all kind of Slice, not only `Slice(UInt8)`.. But maybe it's ok? or maybe we can add a macro check inside the `to_<num_type>` methods that raise when `T != UInt8` ?
Another possibility would be to make `Bytes` a new type (not just an alias) that inherits from `Slice(UInt8)` so it's easier to add methods that works specifically on bytes.

WDYT?

**Comment by RX14** [comment]
I agree that `String.unsafe_from_slice` is probably a better solution. 

Unfortunately, it can't be implemented because strings are layed out differently from every other class. Strings don't have a pointer to their data, they have their data directly after their class in memory, so it's impossible to turn a Bytes into a String without allocating.

**Comment by bew** [comment]
Having `to_i` on `Slice` is only logic for `Slice(UInt8)`, right @asterite ? What do you think should be done for a `Slice(Person)` ?

**Comment by asterite** [comment]
Give an error. We already have methods on slice that only work for byte slices. And byte slices are the most common slice type (in fact, but I would happily change Slice(T) to a non generic Bytes type) and remove the Slice type from the std, or maybe keep the two but return Bytes instead of Slice(UInt8) in all cases).

**Comment by bew** [comment]
Ok, that's what I thought too. And I was also thinking about a `Bytes` type which inherits from `Slice(UInt8)` and would have Bytes-specific methods like `to_*` methods.

**Comment by asterite** [comment]
Can't inherit from struct. They would be two different types with a few similar methods.

**Comment by bew** [comment]
oh right, then maybe make it as a wrapper like:
```cr
struct Bytes
  def initialize(@raw : Slice(UInt8))
  end

  forward_missing_to @raw

  def to_i(...)
    # ...
  end
end
```

**Comment by asterite** [comment]
A little copying is better than a little dependency. I won't get tired of saying that.

**Comment by RX14** [comment]
just put it on string

**Comment by bew** [comment]
@asterite by "a little copying" you mean the 550 lines of `Slice` ?

**Comment by straight-shoota** [comment]
I guess this could be a nice use case for extension methods as described in #6438

**Comment by RX14** [comment]
it's easy enough to throw a `{% raise unless T == UInt8 %}` in the aforementioned methods and you'll hardly be able to tell the difference

**Comment by jhass** [comment]
@bew A random example would be [BCD](https://en.wikipedia.org/wiki/Binary-coded_decimal)

**Comment by straight-shoota** [comment]
I would very much like to make basic string processing methods available for plain byte slices.
This would certainly help for some low level implementations to avoid unnecessary string allocations.

Maybe an option for this could be the introduction of a special type (perhaps `StringBuffer`?). It would wrap a `Bytes` instance and provide additional string related functionality.[^1]
This type clearly dedicates it to hold string contents. It makes string related methods available.
Another benefit is that it can be used in other places as a replacement for `String` with string-like semantics. For example if a `StringBuffer` and `String` with the same contents compare equal, it can be used for hash lookups on string keys, without allocating the search string (ref https://github.com/crystal-lang/crystal/pull/13844#user-content-fn-1-28ca2fcec204f62bf160b98e50402de5).

Example:
```cr
record StringBuffer, data : Bytes

class String
  def ==(other : StringBuffer)
    to_slice == other.data
  end
end

hash = {"foo" => "bar"}
hash[StringBuffer.new("xfoox".to_slice[1..-2])] # => "bar"
```

[^1]: If it was possible to inherit from `Slice`, it would be a subtype of `Slice(UInt8)`.

**Comment by straight-shoota** [comment]
A relevant consideration when working with arbitrary byte slices interpreted as string content is that there's no guarantee for a trailing null byte.
The payload of `String` *always* ends with a null byte. Many algorithms implicitly rely on that when reading through a string. For example, a loop consuming whitespace will automatically break on a null byte. There's no need for explicit bounds checking.

With arbitrary byte slices, we cannot be sure that there's a trailing null byte. And in most practical use cases, there won't be any.
So algorithms working on a byte slice must employ explicit bounds checking.

**Comment by straight-shoota** [comment]
I don't follow what difference class methods would make in this context compared to a typed instance with instance methods?

**Comment by HertzDevil** [comment]
Having `StringBuffer#to_i` instead of `StringBuffer.to_i(Bytes)` means having the possibility of storing the `StringBuffer` as part of another object, hence prolonging the original string's lifetime, which we should avoid (especially after working with related memory issues recently).

**Comment by straight-shoota** [comment]
Isn't it the same with storing `Bytes` that is passed to `.to_i`? I don't see any difference whether the type that can potentially be stored is `StringBuffer` or `Bytes`. The result would be the same.

I wouldn't worry about this too much though either way. We're talking about a very low level instrument intended for optimizing heavily used algorithms. It should be fine to require some careful handling in this context.
